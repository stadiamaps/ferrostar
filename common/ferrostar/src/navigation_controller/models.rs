//! State and configuration data models.

use crate::deviation_detection::{RouteDeviation, RouteDeviationTracking};
use crate::models::{RouteStep, SpokenInstruction, UserLocation, VisualInstruction, Waypoint};
#[cfg(feature = "alloc")]
use alloc::vec::Vec;
use geo::LineString;
#[cfg(feature = "wasm-bindgen")]
use serde::{Deserialize, Serialize};

/// High-level state describing progress through a route.
#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
#[cfg_attr(feature = "wasm-bindgen", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "wasm-bindgen", serde(rename_all = "camelCase"))]
pub struct TripProgress {
    /// The distance to the next maneuver, in meters.
    pub distance_to_next_maneuver: f64,
    /// The total distance remaining in the trip, in meters.
    ///
    /// This is the sum of the distance remaining in the current step and the distance remaining in all subsequent steps.
    pub distance_remaining: f64,
    /// The total duration remaining in the trip, in seconds.
    pub duration_remaining: f64,
}

/// The state of a navigation session.
///
/// This is produced by [`NavigationController`](super::NavigationController) methods
/// including [`get_initial_state`](super::NavigationController::get_initial_state)
/// and [`update_user_location`](super::NavigationController::update_user_location).
#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[cfg_attr(feature = "wasm-bindgen", derive(Serialize, Deserialize))]
pub enum TripState {
    /// The navigation controller is idle and there is no active trip.
    Idle,
    #[cfg_attr(feature = "wasm-bindgen", serde(rename_all = "camelCase"))]
    /// The navigation controller is actively navigating a trip.
    Navigating {
        /// The index of the closest coordinate to the user's snapped location.
        ///
        /// This index is relative to the *current* [`RouteStep`]'s geometry.
        current_step_geometry_index: Option<u64>,
        /// A location on the line string that
        snapped_user_location: UserLocation,
        /// The ordered list of steps that remain in the trip.
        ///
        /// The step at the front of the list is always the current step.
        /// We currently assume that you cannot move backward to a previous step.
        remaining_steps: Vec<RouteStep>,
        /// Remaining waypoints to visit on the route.
        ///
        /// The waypoint at the front of the list is always the *next* waypoint "goal."
        /// Unlike the current step, there is no value in tracking the "current" waypoint,
        /// as the main use of waypoints is recalculation when the user deviates from the route.
        /// (In most use cases, a route will have only two waypoints, but more complex use cases
        /// may have multiple intervening points that are visited along the route.)
        /// This list is updated as the user advances through the route.
        remaining_waypoints: Vec<Waypoint>,
        /// The trip progress includes information that is useful for showing the
        /// user's progress along the full navigation trip, the route and its components.
        progress: TripProgress,
        /// The route deviation status: is the user following the route or not?
        deviation: RouteDeviation,
        /// The visual instruction that should be displayed in the user interface.
        visual_instruction: Option<VisualInstruction>,
        /// The most recent spoken instruction that should be synthesized using TTS.
        ///
        /// Note it is the responsibility of the platform layer to ensure that utterances are not synthesized multiple times. This property simply reports the current spoken instruction.
        spoken_instruction: Option<SpokenInstruction>,
        /// Annotation data at the current location.
        /// This is represented as a json formatted byte array to allow for flexible encoding of custom annotations.
        annotation_json: Option<String>,
    },
    /// The navigation controller has reached the end of the trip.
    Complete,
}

pub enum StepAdvanceStatus {
    /// Navigation has advanced, and the information on the next step is embedded.
    Advanced {
        step: RouteStep,
        linestring: LineString,
    },
    /// Navigation has reached the end of the route.
    EndOfRoute,
}

/// Controls filtering/post-processing of user course by the [`NavigationController`].
#[derive(Debug, Copy, Clone)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[cfg_attr(feature = "wasm-bindgen", derive(Deserialize))]
pub enum CourseFiltering {
    /// Snap the user's course to the current step's linestring using the next index in the step's geometry.
    ///
    // TODO: We could make this more flexible by allowing the user to specify number of indices to average, etc.
    SnapToRoute,

    /// Use the raw course as reported by the location provider with no processing.
    Raw,
}

/// The step advance mode describes when the current maneuver has been successfully completed,
/// and we should advance to the next step.
#[derive(Debug, Copy, Clone)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[cfg_attr(feature = "wasm-bindgen", derive(Deserialize))]
pub enum StepAdvanceMode {
    /// Never advances to the next step automatically;
    /// requires calling [`NavigationController::advance_to_next_step`](super::NavigationController::advance_to_next_step).
    ///
    /// You can use this to implement custom behaviors in external code.
    Manual,
    /// Automatically advances when the user's location is close enough to the end of the step
    #[cfg_attr(feature = "wasm-bindgen", serde(rename_all = "camelCase"))]
    DistanceToEndOfStep {
        /// Distance to the last waypoint in the step, measured in meters, at which to advance.
        distance: u16,
        /// The minimum required horizontal accuracy of the user location, in meters.
        /// Values larger than this cannot trigger a step advance.
        minimum_horizontal_accuracy: u16,
    },
    /// Automatically advances when the user's distance to the *next* step's linestring  is less
    /// than the distance to the current step's linestring.
    #[cfg_attr(feature = "wasm-bindgen", serde(rename_all = "camelCase"))]
    RelativeLineStringDistance {
        /// The minimum required horizontal accuracy of the user location, in meters.
        /// Values larger than this cannot trigger a step advance.
        minimum_horizontal_accuracy: u16,
        /// At this (optional) distance, navigation should advance to the next step regardless
        /// of which `LineString` appears closer.
        automatic_advance_distance: Option<u16>,
    },
}

#[derive(Clone)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
#[cfg_attr(feature = "wasm-bindgen", derive(Deserialize))]
#[cfg_attr(feature = "wasm-bindgen", serde(rename_all = "camelCase"))]
pub struct NavigationControllerConfig {
    /// Configures when navigation advances to the next step in the route.
    pub step_advance: StepAdvanceMode,
    /// Configures when the user is deemed to be off course.
    ///
    /// NOTE: This is distinct from the action that is taken.
    /// It is only the determination that the user has deviated from the expected route.
    pub route_deviation_tracking: RouteDeviationTracking,
    /// Configures how the heading component of the snapped location is reported in [`TripState`].
    pub snapped_location_course_filtering: CourseFiltering,
}
