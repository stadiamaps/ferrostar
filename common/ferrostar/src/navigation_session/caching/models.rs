use serde::{Deserialize, Serialize};

#[cfg(all(feature = "uniffi", not(feature = "wasm_js")))]
use crate::{models::Route, navigation_controller::models::TripState, UtcDateTime};

#[cfg(all(feature = "uniffi", not(feature = "wasm_js")))]
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
pub struct NavigationCachingConfig {
    pub cache_interval_seconds: i64,
    pub max_age_seconds: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg(all(feature = "uniffi", not(feature = "wasm_js")))]
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
pub struct NavigationSessionSnapshot {
    pub saved_at: UtcDateTime,
    pub route: Route,
    pub trip_state: Option<TripState>,
}
