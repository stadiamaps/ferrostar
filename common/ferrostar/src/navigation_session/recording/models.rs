use chrono::Utc;
use serde::{Deserialize, Serialize};

use crate::{
    models::Route,
    navigation_controller::{
        models::{
            NavigationControllerConfig, SerializableNavState,
            SerializableNavigationControllerConfig, TripState,
        },
        step_advance::SerializableStepAdvanceCondition,
    },
};

#[cfg(feature = "wasm-bindgen")]
use tsify::Tsify;

/// A builder for serializing a navigation recording.
#[derive(Serialize, Deserialize, Clone)]
pub struct NavigationRecordingMetadata {
    /// Version of Ferrostar used.
    pub version: String,
    /// Initial timestamp of the recording.
    pub initial_timestamp: i64,
    /// Configuration of the navigation controller.
    pub config: SerializableNavigationControllerConfig,
    /// Initial route used during navigation.
    pub initial_route: Route,
}

impl NavigationRecordingMetadata {
    /// Creates a new navigation recorder with route configuration and initial state.
    pub fn new(config: NavigationControllerConfig, initial_route: Route) -> Self {
        Self {
            version: env!("CARGO_PKG_VERSION").to_string(),
            initial_timestamp: Utc::now().timestamp_millis(),
            config: SerializableNavigationControllerConfig::from(config),
            initial_route,
        }
    }

    /// Serializes the navigation recording to a JSON string.
    pub fn to_json(&self, events: Vec<NavigationRecordingEvent>) -> Result<String, RecordingError> {
        let recording = NavigationRecording {
            recording: self.clone(),
            events,
        };
        serde_json::to_string(&recording).map_err(|e| RecordingError::SerializationError {
            error: e.to_string(),
        })
    }
}

/// A navigation session recording.
///
/// Internally this contains the full event stream.
// TODO: Hints for how you would typically use / interact with this? You can link to other types (and functions) by the way like [`NavigationReplay`] :)  and a list of navigation events.
#[derive(Serialize, Deserialize, Clone)]
pub struct NavigationRecording {
    #[serde(flatten)]
    pub recording: NavigationRecordingMetadata,
    /// List of navigation events in order of occurrence.
    pub events: Vec<NavigationRecordingEvent>,
}

impl NavigationRecording {
    /// Deserializes a previously saved navigation recording from a JSON string.
    pub fn from_json(json: &str) -> Self {
        serde_json::from_str(json)
            .map_err(|e| RecordingError::SerializationError {
                error: e.to_string(),
            })
            .unwrap()
    }
}

/// A session recording error.
#[derive(Debug, Serialize)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Error))]
#[cfg_attr(feature = "std", derive(thiserror::Error))]
pub enum RecordingError {
    /// Error during serialization.
    #[cfg_attr(
        feature = "std",
        error("Error serializing navigation recording: {error}.")
    )]
    SerializationError { error: String },

    /// Recording is not enabled for this controller.
    #[cfg_attr(
        feature = "std",
        error("Recording is not enabled for this controller.")
    )]
    RecordingNotEnabled,
}

/// An event that occurs during navigation.
///
/// This is used for the optional session recording / telemetry.
#[derive(Clone, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
#[cfg_attr(feature = "wasm-bindgen", derive(Tsify))]
pub struct NavigationRecordingEvent {
    /// The timestamp of the event in milliseconds since Jan 1, 1970 UTC.
    pub timestamp: i64,
    /// Data associated with the event.
    pub event_data: NavigationRecordingEventData,
}

impl NavigationRecordingEvent {
    pub fn new(event_data: NavigationRecordingEventData) -> Self {
        Self {
            timestamp: Utc::now().timestamp_millis(),
            event_data,
        }
    }

    /// Create a [`NavigationRecordingEventData::StateUpdate`] event from a [`SerializableNavState`]
    pub fn state_update(serializable_nav_state: SerializableNavState) -> Self {
        Self::new(NavigationRecordingEventData::StateUpdate {
            trip_state: serializable_nav_state.trip_state,
            step_advance_condition: serializable_nav_state.step_advance_condition,
        })
    }

    pub fn timestamp(&self) -> i64 {
        self.timestamp
    }
}

/// The event type.
///
/// For full replayability, we record things like rerouting, and not just location updates.
#[derive(Clone, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[cfg_attr(feature = "wasm-bindgen", derive(Tsify))]
pub enum NavigationRecordingEventData {
    StateUpdate {
        trip_state: TripState,
        step_advance_condition: SerializableStepAdvanceCondition,
    },
    // TODO: Figure out how to record re-routes.
    RouteUpdate {
        /// Updated route.
        route: Route,
    },
}
