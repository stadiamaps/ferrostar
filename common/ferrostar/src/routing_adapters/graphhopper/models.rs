use serde::{Deserialize};
use alloc::collections::BTreeMap as HashMap;
use crate::models::{ GeographicCoordinate };

#[derive(Deserialize, Debug)]
pub struct GraphHopperRouteResponse {
    pub paths: Vec<GraphHopperPath>,
    pub message: Option<String>
}

#[derive(Deserialize, Debug)]
#[serde(untagged)]
pub enum DetailEntryValue {
    Int(i32),
    Str(String),
    Bool(bool),
    Float(f64),
}

#[derive(Deserialize, Debug)]
#[serde(from = "Vec<serde_json::Value>")]
pub struct DetailEntry {
    pub start_index: usize,
    pub end_index: usize,
    pub value: Option<DetailEntryValue>,
}

impl From<Vec<serde_json::Value>> for DetailEntry {
    fn from(vec: Vec<serde_json::Value>) -> Self {
        let start_index = vec[0].as_i64().unwrap_or_default() as usize;
        let end_index = vec[1].as_i64().unwrap_or_default() as usize;

        let value = match &vec[2] {
            serde_json::Value::String(s) => Some(DetailEntryValue::Str(s.clone())),
            serde_json::Value::Number(n) => {
                if n.is_i64() {
                    Some(DetailEntryValue::Int(n.as_i64().unwrap() as i32))
                } else {
                    Some(DetailEntryValue::Float(n.as_f64().unwrap()))
                }
            },
            serde_json::Value::Bool(b) => Some(DetailEntryValue::Bool(*b)),
            serde_json::Value::Null => None,
            _ => None,
        };

        DetailEntry { start_index, end_index, value }
    }
}

#[derive(Deserialize, Debug)]
pub struct GraphHopperPath {
    pub distance: f64,
    pub time: f64,
    pub bbox: Vec<f64>,
    pub instructions: Vec<GraphHopperInstruction>,
    pub details: HashMap<String, Vec<DetailEntry>>,
    pub points: String, // encoded polyline
    pub points_encoded:	bool,
    pub points_encoded_multiplier: f64,
}

#[derive(Deserialize, Debug)]
pub struct GraphHopperInstruction {
    pub distance: f64,
    pub time: f64,
    pub heading: Option<f64>,
    pub exit_number: Option<u32>,
    pub turn_angle: Option<f64>,
    pub sign: i32,
    pub interval: Vec<usize>,
    pub text: String,
    pub street_ref: Option<String>,
    pub street_name: String,
}

// temporary structure until we understand internal maxspeed handling
pub struct MaxSpeedEntry {
    pub geometry: Vec<GeographicCoordinate>,
    pub speed_limit: Option<f64>,
    pub unit: String,
}
