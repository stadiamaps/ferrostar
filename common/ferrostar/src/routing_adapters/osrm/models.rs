//! OSRM models from the API spec: <http://project-osrm.org/docs/v5.5.1/api/>
//!
//! Note that in some cases we optionally allow for extensions that have been made to the spec
//! by others which are now pseudo-standardized (ex: Mapbox). We omit some fields which are not
//! needed for navigation.

use crate::models::{ManeuverModifier, ManeuverType};
use alloc::{string::String, vec::Vec};
use serde::Deserialize;
use serde_json::Value;
#[cfg(feature = "alloc")]
use std::collections::HashMap;

#[derive(Deserialize, Debug)]
#[serde(transparent)]
pub struct Coordinate {
    tuple: (f64, f64),
}

impl Coordinate {
    pub fn latitude(&self) -> f64 {
        self.tuple.1
    }

    pub fn longitude(&self) -> f64 {
        self.tuple.0
    }
}

#[derive(Deserialize, Debug)]
pub struct RouteResponse {
    /// The response code.
    ///
    /// Ok indicates success. TODO: enumerate others?
    pub code: String,
    pub routes: Vec<Route>,
    pub waypoints: Vec<Waypoint>,
}

/// A route between two or more waypoints.
#[derive(Deserialize, Debug)]
pub struct Route {
    /// The estimated travel time, in seconds.
    pub duration: f64,
    /// The distance traveled by the route, in meters.
    pub distance: f64,
    /// The geometry of the route.
    ///
    /// NOTE: This library assumes that 1) an overview geometry will always be requested, and
    /// 2) that it will be a polyline (whether it is a polyline5 or polyline6 can be determined
    /// by the [`crate::routing_adapters::RouteResponseParser`]).
    pub geometry: String,
    /// The legs between the given waypoints.
    pub legs: Vec<RouteLeg>,
}

/// A route between exactly two waypoints.
#[derive(Deserialize, Debug)]
pub struct RouteLeg {
    pub annotation: Option<AnyAnnotation>,
    /// The estimated travel time, in seconds.
    pub duration: f64,
    /// The distance traveled this leg, in meters.
    pub distance: f64,
    /// A sequence of steps with turn-by-turn instructions.
    pub steps: Vec<RouteStep>,
    /// A Mapbox and Valhalla extension which indicates which waypoints are passed through rather than creating a new leg.
    #[serde(default)]
    pub via_waypoints: Vec<ViaWaypoint>,
}

#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct AnyAnnotation {
    #[serde(flatten)]
    pub values: HashMap<String, Vec<Value>>,
}

/// The max speed json units that can be associated with annotations.
/// [MaxSpeed/Units](https://wiki.openstreetmap.org/wiki/Map_Features/Units#Speed)
/// TODO: We could generalize this as or map from this to a `UnitSpeed` enum.
#[derive(Deserialize, PartialEq, Debug, Clone)]
pub enum MaxSpeedUnits {
    #[serde(rename = "km/h")]
    KilometersPerHour,
    #[serde(rename = "mph")]
    MilesPerHour,
}

/// The local posted speed limit between a pair of coordinates.
#[derive(Debug, Clone)]
#[allow(dead_code)] // TODO: https://github.com/stadiamaps/ferrostar/issues/271
pub enum MaxSpeed {
    Unknown { unknown: bool },
    Known { speed: f64, unit: MaxSpeedUnits },
}

#[allow(dead_code)] // TODO: https://github.com/stadiamaps/ferrostar/issues/271
impl MaxSpeed {
    /// Get the max speed as meters per second.
    pub fn get_as_meters_per_second(&self) -> Option<f64> {
        match self {
            MaxSpeed::Known { speed, unit } => match unit {
                MaxSpeedUnits::KilometersPerHour => Some(speed * 0.27778),
                MaxSpeedUnits::MilesPerHour => Some(speed * 0.44704),
            },
            #[allow(unused)]
            MaxSpeed::Unknown { unknown } => None,
        }
    }
}

#[derive(Deserialize, Debug)]
pub struct RouteStep {
    /// The distance from the start of the current maneuver to the following step, in meters.
    pub distance: f64,
    /// The estimated travel time, in seconds.
    pub duration: f64,
    /// The (unsimplified) geometry of the route segment.
    ///
    /// NOTE: This library assumes that the geometry will always be a polyline.
    pub geometry: String,
    /// The name of the way along which travel proceeds.
    pub name: Option<String>,
    /// A reference number or code for the way (if one is available).
    #[serde(rename = "ref")]
    pub reference: Option<String>,
    /// A pronunciation hint for the name of the way.
    pub pronunciation: Option<String>,
    /// The mode of transportation.
    pub mode: Option<String>,
    /// The maneuver for this step
    pub maneuver: StepManeuver,
    /// TODO: docs
    pub intersections: Vec<Intersections>,

    /// A list of exits (name or number), separated by semicolons.
    ///
    /// NOTE: This annotation is not in the official spec, but is a common extension used by Mapbox
    /// and Valhalla.
    pub exits: Option<String>,

    /// The side of the way on which traffic proceeds.
    ///
    /// NOTE: This annotation is not in the official spec, but is a common extension used by Mapbox
    /// and Valhalla.
    pub driving_side: Option<String>,
    // Mapbox and Valhalla extensions that might be useful later
    // pub rotary_name: Option<String>,
    // pub rotary_pronunciation: Option<String>,
    /// Textual instructions that are displayed as a banner; supported by Mapbox and Valhalla
    #[serde(default, rename = "bannerInstructions")]
    pub banner_instructions: Vec<BannerInstruction>,
    /// Textual instructions that are displayed as a banner; supported by Mapbox and Stadia Maps
    #[serde(default, rename = "voiceInstructions")]
    pub voice_instructions: Vec<VoiceInstruction>,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct BannerInstruction {
    /// How far (in meters) from the upcoming maneuver the instruction should start being displayed
    pub distance_along_geometry: f64,
    pub primary: BannerContent,
    pub secondary: Option<BannerContent>,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct VoiceInstruction {
    pub announcement: String,
    pub ssml_announcement: Option<String>,
    /// How far (in meters) from the upcoming maneuver the instruction should be announced
    pub distance_along_geometry: f64,
}

#[derive(Deserialize, Debug)]
pub struct BannerContent {
    pub text: String,
    #[serde(rename = "type")]
    pub maneuver_type: Option<ManeuverType>,
    #[serde(rename = "modifier")]
    pub maneuver_modifier: Option<ManeuverModifier>,
    /// The degree at which the maneuver exits the roundabout.
    /// 180 indicates that the effect of exiting will be to continue in the same direction as
    /// the original travel.
    #[serde(rename = "degrees")]
    pub roundabout_exit_degrees: Option<u16>,
}

#[derive(Deserialize, Debug)]
pub struct StepManeuver {
    /// The coordinate at which the maneuver takes place.
    pub location: Coordinate,
    /// The clockwise angle from true north to the direction of travel immediately *before*
    /// the maneuver.
    pub bearing_before: u16,
    /// The clockwise angle from true north to the direction of travel immediately *after*
    /// the maneuver.
    pub bearing_after: u16,
    /// A string indicating the type of maneuver.
    ///
    /// Note that even though there are `new name` and `notification` instructions, the
    /// `mode` and `name` (of the parent [`RouteStep`]) can change between *any* pair of instructions.
    /// They only offer a fallback in case there is nothing else to report.
    /// TODO: Model this as an enum. Note that new types may be introduced, and anything unknown to the client should be handled like a turn.
    #[serde(rename = "type")]
    pub maneuver_type: String,
    /// An optional string indicating the direction change of the maneuver.
    /// TODO: Model this as an enum.
    pub modifier: Option<String>,
    /// Non-standard extension in Mapbox and Valhalla where the instruction is computed server-side
    instruction: Option<String>,
}

impl StepManeuver {
    // TODO: This is a placeholder implementation.
    // Most commercial offerings offer server-side synthesis of voice instructions.
    // However, we might consider synthesizing these locally too.
    // This will be rather cumbersome with localization though.
    fn synthesize_instruction(&self, _locale: &str) -> String {
        String::from("TODO: OSRM instruction synthesis")
    }

    pub fn get_instruction(&self) -> String {
        self.instruction
            .clone()
            .unwrap_or_else(|| self.synthesize_instruction("en-US"))
    }
}

#[derive(Deserialize, Debug)]
pub struct Intersections {
    /// The location of the intersection
    pub location: Coordinate,
    /// A list of bearing values that are available at the intersection.
    ///
    /// These describe all available roads at the intersection.
    pub bearings: Vec<u16>,
    /// An array of strings signifying the classes (as specified in the profile) of the road
    /// exiting the intersection.
    ///
    /// Note that Valhalla servers do not return this property.
    #[serde(default)]
    pub classes: Vec<String>,
    /// A list of entry flags, corresponding 1:1 to the list of bearings.
    ///
    /// This value indicates whether the respective road could be entered on a valid route (not
    /// violating some restriction).
    pub entry: Vec<bool>,
    /// An index into the bearings/entry array used to calculate the bearing just before the turn.
    ///
    /// The clockwise angle from true north to the direction of travel immediately before the
    /// maneuver/passing the intersection.
    /// To get the bearing in the direction of driving, the bearing has to be rotated by
    /// 180 degrees. Not supplied for `depart` maneuvers.
    #[serde(default)]
    #[serde(rename = "in")]
    pub intersection_in: usize,
    /// An index into the bearings/entry array used to calculate the bearing just before the turn.
    ///
    /// The clockwise angle from true north to the direction of travel immediately after the
    /// maneuver/passing the intersection.  Not supplied for `arrive` maneuvers.
    #[serde(default)]
    #[serde(rename = "out")]
    pub intersection_out: usize,
    /// A list of turn [`Lane`]s available at the intersection (if info is available).
    ///
    /// Lanes are listed in left-to-right order.
    #[serde(default)]
    pub lanes: Vec<Lane>,
}

#[derive(Deserialize, Debug)]
pub struct Lane {
    /// An indication (ex: marking on the road, sign, etc.) for a turn lane.
    ///
    /// A lane may have multiple indications (ex: both straight and left),
    /// TODO: Turn this into an enum
    pub indications: Vec<String>,
    /// Whether the lane is a valid choice for the current maneuver
    pub valid: bool,
    // TODO: Mapbox and Valhalla extensions: `active` and `valid_indication`
}

#[derive(Deserialize, Debug)]
pub struct Waypoint {
    /// The name of the street that the waypoint snapped to.
    pub name: Option<String>,
    /// The distance (in meters) between the snapped point and the input coordinate.
    pub distance: Option<f64>,
    /// The waypoint's location on the road network.
    pub location: Coordinate,
}

#[derive(Deserialize, Debug)]
pub struct ViaWaypoint {
    /// The distance (in meters) from the leg origin
    pub distance_from_start: f64,
    /// The geometry point index of the location (leg-specific).
    pub geometry_index: f64,
    /// The waypoint's index in the array of waypoints.
    pub waypoint_index: usize,
}

#[cfg(test)]
mod tests {
    use super::*;

    // TODO: Route
    // TODO: RouteLeg

    #[test]
    fn deserialize_annotation() {
        // Example from Mapbox's public docs, which include several annotations not supported at
        // the time of this writing.
        let data = r#"{
            "distance": [
                4.294596842089401,
                5.051172053200946,
                5.533254065167979,
                6.576513793849532,
                7.4449640160938015,
                8.468757534990829,
                15.202780313562714,
                7.056346577326572
            ],
            "duration": [
                1,
                1.2,
                2,
                1.6,
                1.8,
                2,
                3.6,
                1.7
            ],
            "speed": [
                4.3,
                4.2,
                2.8,
                4.1,
                4.1,
                4.2,
                4.2,
                4.2
            ],
            "congestion": [
                "low",
                "moderate",
                "moderate",
                "moderate",
                "heavy",
                "heavy",
                "heavy",
                "heavy"
            ],
            "maxspeed": [
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              }
            ]
        }"#;

        let annotation: AnyAnnotation =
            serde_json::from_str(data).expect("Failed to parse Annotation");

        insta::with_settings!({sort_maps => true}, {
            insta::assert_yaml_snapshot!(annotation.values);
        });
    }

    #[test]
    fn deserialize_banner_instruction() {
        // Example from Mapbox's public docs
        let data = r#"
        {
          "distanceAlongGeometry": 100,
          "primary": {
            "type": "turn",
            "modifier": "left",
            "text": "I 495 North / I 95",
            "components": [
              {
                "text": "I 495",
                "imageBaseURL": "https://s3.amazonaws.com/mapbox/shields/v3/i-495",
                "type": "icon"
              },
              {
                "text": "North",
                "type": "text",
                "abbr": "N",
                "abbr_priority": 0
              },
              {
                "text": "/",
                "type": "delimiter"
              },
              {
                "text": "I 95",
                "imageBaseURL": "https://s3.amazonaws.com/mapbox/shields/v3/i-95",
                "type": "icon"
              }
            ]
          },
          "secondary": {
            "type": "turn",
            "modifier": "left",
            "text": "Baltimore / Northern Virginia",
            "components": [
              {
                "text": "Baltimore",
                "type": "text"
              },
              {
                "text": "/",
                "type": "text"
              },
              {
                "text": "Northern Virginia",
                "type": "text"
              }
            ]
          },
          "sub": {
            "text": "",
            "components": [
              {
                "text": "",
                "type": "lane",
                "directions": [
                  "left"
                ],
                "active": true
              },
              {
                "text": "",
                "type": "lane",
                "directions": [
                  "left",
                  "straight"
                ],
                "active": true
              },
              {
                "text": "",
                "type": "lane",
                "directions": [
                  "right"
                ],
                "active": false
              }
            ]
          }
        }
        "#;

        let instruction: BannerInstruction =
            serde_json::from_str(data).expect("Failed to parse Annotation");

        assert_eq!(instruction.distance_along_geometry, 100.0);
        assert_eq!(instruction.primary.text, "I 495 North / I 95");
        assert_eq!(instruction.primary.maneuver_type, Some(ManeuverType::Turn));
        assert_eq!(
            instruction.primary.maneuver_modifier,
            Some(ManeuverModifier::Left)
        );

        let secondary = instruction.secondary.expect("Expected secondary content");
        assert_eq!(secondary.text, "Baltimore / Northern Virginia");
        assert_eq!(secondary.maneuver_type, Some(ManeuverType::Turn));
        assert_eq!(secondary.maneuver_modifier, Some(ManeuverModifier::Left));
    }

    #[test]
    fn test_max_speed_unknown() {
        let max_speed = MaxSpeed::Unknown { unknown: true };
        assert_eq!(max_speed.get_as_meters_per_second(), None);
    }

    #[test]
    fn test_max_speed_kph() {
        let max_speed = MaxSpeed::Known {
            speed: 100.0,
            unit: MaxSpeedUnits::KilometersPerHour,
        };
        assert_eq!(
            max_speed.get_as_meters_per_second(),
            Some(27.778000000000002)
        );
    }

    #[test]
    fn test_max_speed_mph() {
        let max_speed = MaxSpeed::Known {
            speed: 60.0,
            unit: MaxSpeedUnits::MilesPerHour,
        };
        assert_eq!(max_speed.get_as_meters_per_second(), Some(26.8224));
    }
}
