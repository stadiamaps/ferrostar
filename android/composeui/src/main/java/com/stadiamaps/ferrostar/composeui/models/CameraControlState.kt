package com.stadiamaps.ferrostar.composeui.models

sealed class CameraControlState {
  data object Hidden : CameraControlState()

  data class ShowRecenter(val updateCamera: () -> Unit) : CameraControlState()

  data class ShowRouteOverview(val updateCamera: () -> Unit) : CameraControlState()
}
