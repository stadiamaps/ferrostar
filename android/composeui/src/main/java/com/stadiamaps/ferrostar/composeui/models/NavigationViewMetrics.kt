package com.stadiamaps.ferrostar.composeui.models

import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.DpSize
import androidx.compose.ui.unit.dp

data class NavigationViewMetrics(
    val progressViewSize: DpSize = DpSize(0.dp, 0.dp),
    val instructionsViewSize: DpSize = DpSize(0.dp, 0.dp),
    val buttonSize: DpSize,
) {

  /**
   * Returns the MapView's safe insets.
   *
   * @param start Optional additional start padding. Default is 0.dp.
   * @param top Optional additional top padding. Default is instructionsViewSize.height.
   * @param end Optional additional end padding. Default is 0.dp.
   * @param bottom Optional additional bottom padding. Default is progressViewSize.height.
   * @return The calculated padding insets.
   */
  fun mapViewInsets(
      start: Dp = 0.dp,
      top: Dp = 0.dp,
      end: Dp = 0.dp,
      bottom: Dp = 0.dp,
  ): PaddingValues {
    return PaddingValues(
        start = start,
        top = instructionsViewSize.height + top,
        end = end,
        bottom = progressViewSize.height + buttonSize.height + bottom,
    )
  }
}
