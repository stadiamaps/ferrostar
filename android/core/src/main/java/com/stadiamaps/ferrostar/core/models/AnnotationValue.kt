package com.stadiamaps.ferrostar.core.models

import com.squareup.moshi.Json

data class AnnotationValue(
  val distance: Double,
  val duration: Double,
  @Json(name = "max_speed_mps")
  val maxSpeedMps: Double?,
  @Json(name = "speed_mps")
  val speedMps: Double
)