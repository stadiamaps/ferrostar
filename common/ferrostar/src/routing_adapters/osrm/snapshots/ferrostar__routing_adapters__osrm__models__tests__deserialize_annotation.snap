---
source: ferrostar/src/routing_adapters/osrm/models.rs
expression: annotation.values
---
congestion:
  - low
  - moderate
  - moderate
  - moderate
  - heavy
  - heavy
  - heavy
  - heavy
distance:
  - 4.294596842089401
  - 5.051172053200946
  - 5.533254065167979
  - 6.576513793849532
  - 7.4449640160938015
  - 8.468757534990829
  - 15.202780313562714
  - 7.056346577326572
duration:
  - 1
  - 1.2
  - 2
  - 1.6
  - 1.8
  - 2
  - 3.6
  - 1.7
maxspeed:
  - speed: 56
    unit: km/h
  - speed: 56
    unit: km/h
  - speed: 56
    unit: km/h
  - speed: 56
    unit: km/h
  - speed: 56
    unit: km/h
  - speed: 56
    unit: km/h
  - speed: 56
    unit: km/h
  - speed: 56
    unit: km/h
speed:
  - 4.3
  - 4.2
  - 2.8
  - 4.1
  - 4.1
  - 4.2
  - 4.2
  - 4.2
